### PR TITLE
Issue 1069: Provide a setting in consumer configuration to specify where to start consuming messages 

### PR DESCRIPTION
**Pull request by XiaoZYang**
### Motivation

This is a fix try for [Issue 1069](https://github.com/apache/incubator-pulsar/issues/1069), which expects a setting in consumer configuration to specify where to start consuming messages.

### Modifications
1. add configuration option *initializeSubscriptionOnLatest* to class ConsumerConfiguration
2. add an overload method for `asyncOpenCursor` in `ManageLedger` to implement different initalization position of the cursor created.
3. modify corresponding testing class
### Result

When a user subscribes a topic the first time, the broker will initialize the cursor the subscription on latest or earliest, which depends on the configuration option  "initializeSubscriptionOnLatest" in class ConsumerConfiguration.

For more details, if user set the option "initializeSubscriptionOnLatest" to "false" when creating a new subscription the cursor will be set at the earliest position, otherwise (user doesn't set it explicitly -- use the default value or user set it to "true") at the latest position.

### CODE REVIEW

File: managed-ledger/src/main/java/org/apache/bookkeeper/mledger/ManagedLedger.java [hunk 1]
```diff
@@ -138,6 +138,18 @@
      */
     public ManagedCursor openCursor(String name) throws InterruptedException, ManagedLedgerException;
 
+    /**
+     * Open a ManagedCursor in this ManagedLedger.
+     * <p>
+     * If the cursors doesn't exist, a new one will be created and its position will be at the end of the ManagedLedger.
```
Comment (zhaijack) [inline managed-ledger/src/main/java/org/apache/bookkeeper/mledger/ManagedLedger.java]: nit: its "default" position will be at the end of .....

File: managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerImpl.java [hunk 2]
```diff
@@ -2026,6 +2035,23 @@ PositionImpl getMarkDeletePositionOfSlowestConsumer() {
         return Pair.create(pos, count);
     }
 
+        /**
```
Comment (zhaijack) [inline managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerImpl.java]: nit: space alignment.

File: pulsar-broker/src/test/java/org/apache/pulsar/broker/service/PersistentMessageFinderTest.java [hunk 3]
```diff
@@ -115,7 +115,7 @@ void testPersistentMessageFinder() throws Exception {
         config.setMaxEntriesPerLedger(entriesPerLedger);
         config.setRetentionTime(1, TimeUnit.HOURS);
         ManagedLedger ledger = factory.open(ledgerAndCursorName, config);
-        ManagedCursorImpl c1 = (ManagedCursorImpl) ledger.openCursor(ledgerAndCursorName);
```
Comment (zhaijack) [inline pulsar-broker/src/test/java/org/apache/pulsar/broker/service/PersistentMessageFinderTest.java]: Seems the implementation of `openCursor(ledgerAndCursorName)` will default call `openCursor(ledgerAndCursorName, true)`, why do we need this change .

File: managed-ledger/src/main/java/org/apache/bookkeeper/mledger/ManagedLedger.java (hunk 1 above)
Comment (XiaoZYang) [reply managed-ledger/src/main/java/org/apache/bookkeeper/mledger/ManagedLedger.java]: that's better, I'll add it

File: pulsar-broker/src/test/java/org/apache/pulsar/broker/service/PersistentMessageFinderTest.java (hunk 3 above)
Comment (XiaoZYang) [reply pulsar-broker/src/test/java/org/apache/pulsar/broker/service/PersistentMessageFinderTest.java]: In previous version code, we initialize the cursor at latest position, so the method `openCursor(ledgerAndCursorName)` has the same behavior with current method `openCursor(ledgerAndCursorName, true)`. Therefore, the new method `openCursor(ledgerAndCursorName )` is a compatible version of former api.
--

File: managed-ledger/src/main/java/org/apache/bookkeeper/mledger/ManagedLedger.java (hunk 1 above)
Comment (XiaoZYang) [reply managed-ledger/src/main/java/org/apache/bookkeeper/mledger/ManagedLedger.java]: I've already added the description in the newest commit. THX for your advice.

File: managed-ledger/src/main/java/org/apache/bookkeeper/mledger/ManagedLedger.java [hunk 4]
```diff
@@ -138,6 +138,20 @@
...
+     *
+     * @param name
+     *            the name associated with the ManagedCursor
+     * @param initializeOnLatest
+     *            the flag tell the method wthether it should intialize the cursor at latest position or not.
+     * @return the ManagedCursor
+     * @throws ManagedLedgerException
+     */
+    public ManagedCursor openCursor(String name, boolean initializeOnLatest) throws InterruptedException, ManagedLedgerException;
```
Comment (merlimat) [inline managed-ledger/src/main/java/org/apache/bookkeeper/mledger/ManagedLedger.java]: rather than using a boolean I'd prefer having an enum here. Reading code with: 

```java
managedLedger.openCursor("name", true);
```

doesn't convey what the `true` means.

```java
managedLedger.openCursor("name", CursorPosition.InitializeOnLatest);
managedLedger.openCursor("name", CursorPosition.InitializeOnEarliest);
```

File: pulsar-client/src/main/java/org/apache/pulsar/client/api/ConsumerConfiguration.java [hunk 5]
```diff
@@ -321,4 +322,21 @@ public ConsumerConfiguration setProperties(Map<String, String> properties) {
     public Map<String, String> getProperties() {
         return properties;
     }
+    
+     /** 
+     * @param initializeSubscriptionOnLatest the initializeSubscriptionOnLatest to set
+     * Set cursor position when subscribing to the topic first time
+     * <p>
+     * Default is {@value true} which means {@link MessageId.lastest}
```
Comment (merlimat) [inline pulsar-client/src/main/java/org/apache/pulsar/client/api/ConsumerConfiguration.java]: It's a bit unclear from this API where the subscription will be initialized if it's not on latest. We should make it clear that it's either "latest" or "earliest" (with latest as default). 

Again, I think that an enum would be a better option here to better convey the 2 options.

File: pulsar-common/src/main/proto/PulsarApi.proto [hunk 6]
```diff
@@ -193,6 +193,9 @@ message CommandSubscribe {
     repeated KeyValue metadata = 10;
 
     optional bool read_compacted = 11;
+	// Signal wthether the subscription will initialize on latest
+	// or not -- earliest
+	optional bool initialize_on_latest = 12 [default = true];
```
Comment (merlimat) [inline pulsar-common/src/main/proto/PulsarApi.proto]: There is already `start_message_id` field, which is currently only used by the `Reader` implementation. We should use that (passing either `MessageId.earliest` or `MessageId.latest`) when subscribing.

File: managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerImpl.java [hunk 7]
```diff
@@ -2026,6 +2035,23 @@ PositionImpl getMarkDeletePositionOfSlowestConsumer() {
...
+     /**
+     * Get the first position written in the managed ledger, alongside with the associated counter
+     */
+    Pair<PositionImpl, Long> getFirstPositionAndCounter() {
+        PositionImpl pos;
+        long count;
+
+        do {
+            pos = getFirstPosition();
+            count = ENTRIES_ADDED_COUNTER_UPDATER.get(this);
```
Comment (merlimat) [inline managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerImpl.java]: The backlog count for this subscription will be wrong here. 

The counter is associated with the last entry written, not with the first one (at least not directly). 

The correct way to initialize it here would be to: 
 1. get first position
 2. get last position and counter (atomically), though the existing method
 3. get number of entries between first and last position (there's method for that)
 4. subtract difference from the last position counter value

File: managed-ledger/src/main/java/org/apache/bookkeeper/mledger/ManagedLedger.java (hunk 4 above)
Comment (merlimat) [reply managed-ledger/src/main/java/org/apache/bookkeeper/mledger/ManagedLedger.java]: Another option here is to reuse  `Position` class: 

eg: 
```java
managedLedger.openCursor("name", Position.earliest);
```

Comment (XiaoZYang) [reply managed-ledger/src/main/java/org/apache/bookkeeper/mledger/ManagedLedger.java]: Thanks for spending on reviewing works, I'll modify the code and apply the enum implementation. 👍

File: pulsar-common/src/main/proto/PulsarApi.proto (hunk 6 above)
Comment (XiaoZYang) [reply pulsar-common/src/main/proto/PulsarApi.proto]: As far as I know about pulsar, the nonpersistent topic subscription (NonpersistentTopicSubscription) use the `start_message_id` field when subscribing. And the persistent topic has no business with it. Therefore, I didn't use that. Do I misunderstand it?

PS. I hope that I don't use nonproper expressions and offend you since I'm not a native English speaker.

File: managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerImpl.java (hunk 7 above)
Comment (XiaoZYang) [reply managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerImpl.java]: Would you mind providing more details about the `counter` here? Thank you. 👍

File: pulsar-common/src/main/proto/PulsarApi.proto [hunk 8]
```diff
@@ -193,6 +193,13 @@ message CommandSubscribe {
     repeated KeyValue metadata = 10;
 
     optional bool read_compacted = 11;
+	// Signal wthether the subscription will initialize on latest
+	// or not -- earliest
+    enum InitialPosition {
```
Comment (merlimat) [inline pulsar-common/src/main/proto/PulsarApi.proto]: formatting is a bit off, it should be 4 spaces indent.

File: managed-ledger/src/main/java/org/apache/bookkeeper/mledger/ManagedLedger.java [hunk 9]
```diff
@@ -193,6 +208,22 @@
...
+     *            the name associated with the ManagedCursor
+     * @param callback
+     *            callback object
+     * @param ctx
+     *            opaque context
+     * @param position
+     *            the cursor will be initialized at lastest position or not when first created
+     *            default is <b>InitialPosition.Latest</b>
+     */
+    public void asyncOpenCursor(String name, OpenCursorCallback callback, Object ctx, InitialPosition initialPosition);
```
Comment (merlimat) [inline managed-ledger/src/main/java/org/apache/bookkeeper/mledger/ManagedLedger.java]: For the `async*`  methods we always use the convention of leaving `callback` and `ctx` as the last parameters for the method.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/MessageDeduplication.java [hunk 10]
```diff
@@ -42,6 +42,7 @@
 import org.apache.pulsar.broker.service.Topic.PublishContext;
 import org.apache.pulsar.common.api.Commands;
 import org.apache.pulsar.common.api.proto.PulsarApi.MessageMetadata;
+import org.apache.pulsar.common.api.proto.PulsarApi.CommandSubscribe.InitialPosition;
```
[the remark this answers is not in the document]
Comment (XiaoZYang) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/MessageDeduplication.java]: I will remove it.

File: managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerImpl.java [hunk 11]
```diff
@@ -2026,6 +2036,29 @@ PositionImpl getMarkDeletePositionOfSlowestConsumer() {
...
+
+            lastPositionAndCounter = getLastPositionAndCounter();
+            entries = getNumberOfEntries();
+            count = lastPositionAndCounter.second - entries;
+            
+            // Ensure no entry was written while reading the two values
+        } while (pos.compareTo(getFirstPosition()) != 0 
+            && lastPositionAndCounter.first.compareTo(getLastPosition()) != 0);
+
+        return Pair.create(pos, count);
```
[the remark this answers is not in the document]
Comment (XiaoZYang) [reply managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerImpl.java]: thanks for your comments, corresponding unit tests will be added soon.

File: managed-ledger/src/main/java/org/apache/bookkeeper/mledger/ManagedLedger.java (hunk 9 above)
Comment (XiaoZYang) [reply managed-ledger/src/main/java/org/apache/bookkeeper/mledger/ManagedLedger.java]: Thx for this suggestion, I will apply that rule.

**Comment by XiaoZYang** [comment]
ping @merlimat , please review this new commit at your convenience.

File: pulsar-client-tools/src/test/java/org/apache/pulsar/client/cli/PulsarClientToolTest.java [hunk 12]
```diff
@@ -51,7 +51,7 @@ public void cleanup() throws Exception {
         super.internalCleanup();
     }
 
-    @Test(timeOut = 10000)
+    // @Test(timeOut = 10000)
```
Comment (merlimat) [inline pulsar-client-tools/src/test/java/org/apache/pulsar/client/cli/PulsarClientToolTest.java]: This test should not be removed

File: pulsar-client/src/main/java/org/apache/pulsar/client/api/ConsumerConfiguration.java [hunk 13]
```diff
@@ -339,4 +342,21 @@ public ConsumerConfiguration setProperties(Map<String, String> properties) {
     public ConsumerConfigurationData getConfigurationData() {
         return conf;
     }
+    
+     /** 
+     * @param subscriptionInitialPosition the initial position at which to set
+     * set cursor  when subscribing to the topic first time
+     * Default is {@value InitialPosition.Latest}
+     */
+    public ConsumerConfiguration setSubscriptionInitialPosition(InitialPosition subscriptionInitialPosition) {
```
Comment (merlimat) [inline pulsar-client/src/main/java/org/apache/pulsar/client/api/ConsumerConfiguration.java]: `ConsumerConfiguration` class has already been set as deprecated for 2.0 release. The new setting is only needed in the new `ConsumerBuilder` interface.

File: pulsar-client/src/main/java/org/apache/pulsar/client/api/ConsumerConfiguration.java [hunk 14]
```diff
@@ -26,6 +26,9 @@
 import java.util.concurrent.TimeUnit;
 
 import org.apache.pulsar.client.impl.conf.ConsumerConfigurationData;
+import org.apache.pulsar.common.api.proto.PulsarApi.CommandSubscribe.InitialPosition;
+import org.apache.pulsar.common.api.proto.PulsarApi.CommandSubscribe.InitialPosition;
+import org.apache.pulsar.common.api.proto.PulsarApi.CommandSubscribe.InitialPosition;
```
Comment (merlimat) [inline pulsar-client/src/main/java/org/apache/pulsar/client/api/ConsumerConfiguration.java]: In any case, we should have a different enum for `InitialPosition` for the user facing API, rather that exposing the `PulsarApi` which is for wire-protocol.

File: faq.md [hunk 15]
```diff
@@ -1,273 +0,0 @@
-# Frequently Asked Questions
```
Comment (merlimat) [inline faq.md]: This file removal appears to be unrelated to this PR.

File: managed-ledger/src/test/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerTest.java [hunk 16]
```diff
@@ -2152,4 +2153,32 @@ public ByteBuf getMessageWithMetadata(byte[] data) throws IOException {
...
+
+        // Since getReadPosition returns the next position, we decrease the entryId by 1
+        PositionImpl p1 = (PositionImpl) cursor1.getReadPosition();
+        PositionImpl p2 = (PositionImpl) cursor2.getReadPosition();
+        
+        Position latest = new PositionImpl(p1.getLedgerId(), p1.getEntryId() - 1);
+        Position earliest = new PositionImpl(p2.getLedgerId(), p2.getEntryId() - 1);
+        assertEquals(latest, ledger.getLastPosition());
+        assertEquals(earliest, ledger.getFirstPosition());
+        ledger.close();
```
Comment (merlimat) [inline managed-ledger/src/test/java/org/apache/bookkeeper/mledger/impl/ManagedLedgerTest.java]: This test is not validating the above counters that are used to track the backlog on the cursor. Eg: `c1.getNumEntriesInBacklog()` or `c1.getNumEntriesToRead()`. 

Also we should test it after a few ledger rollover as well. In most tests, we tune done the max entries per ledgers to a very small number (eg: 2) to exercise those rollover code path.

**Comment by XiaoZYang** [comment]
ping @merlimat 
Since the earliest position should be the first message written to a topic, there is no message is before it, and the counter here means the `ConsumedCount` which equals `all messages - messages in backlog`, so the counter can be set to 0 directly.  Do I misunderstand this?

**Comment by zhaijack** [comment]
@XiaoZYang, Regarding the earliest position, there may be messages before it, but has been deleted. 
Seems counter in getLastPositionAndCounter represent all added entries to this ledger, backlog == added_entry_counter - consumed_counter. In initializeCursorPosition(),  we set `messagesConsumedCounter = lastPositionCounter.second` so the backlog is 0. 

While in this `getFirstPositionAndCounter`, we could make messagesConsumedCounter as value
`ENTRIES_ADDED_COUNTER_UPDATER.get(this)  - getNumberOfEntries(Range.openClosed(firstPosition, lastPosition))`.

Besides this, seems this need rebase, and fix the unit-test error. Thanks.

**Comment by XiaoZYang** [comment]
@zhaijack @merlimat please review this change

**Comment by zhaijack** [comment]
Oh, @XiaoZYang Seems your rebase brings in too much un-related commits and changes, the changed files number is 472 now. And not sure if the CI failure is related.  If it is hard to do the rebase, you could close this PR and re-open a new one with related patch.
